styx-scheduler-service/src/test/java/com/spotify/styx/state/EventFeederTest.java

```diff
@@ -0,0 +1,126 @@
+import com.spotify.styx.util.IsClosed;
+import java.util.List;
+import java.util.function.Consumer;
+import org.junit.Test;
+
+public class EventFeederTest {
+  private final static WorkflowInstance wfi = WorkflowInstance.create(
+      WorkflowId.create("comp1", "work1"), "2017-01-01");
+  private final static SequenceEvent firstEvent = SequenceEvent.create(
+      Event.triggerExecution(wfi, Trigger.natural()), 0, 0);
+  private final static SequenceEvent secondEvent = SequenceEvent.create(
+      Event.dequeue(wfi), 0, 0);
+
+  private List<SequenceEvent> trackedEvents = Lists.newArrayList();
+  private EventFeeder<SequenceEvent> eventFeeder =
+      new EventFeeder<>(new InjectingEventConsumer());
+
+  @Test
+  public void shouldConsumeEvent() throws Exception {
+    eventFeeder.enqueue(firstEvent);
+    await().atMost(5, SECONDS).until(() -> trackedEvents.get(0) != null);
+    assertThat(trackedEvents.get(0), is(firstEvent));
+  }
+
+  @Test
+  public void shouldSkipEventIfExceptionFromEventConsumer() throws Exception {
+    EventFeeder<SequenceEvent> eventConsumer =
+        new EventFeeder<>(new ExceptionalEventConsumer());
+
+    eventConsumer.enqueue(firstEvent);
+    waitAtMost(5, SECONDS).until(() -> eventConsumer.queueSize() == 0);
+    assertThat(trackedEvents.size(), is(0));
+  }
+
+  @Test
+  public void shouldConsumeMoreEvents() throws Exception {
+    eventFeeder.enqueue(firstEvent);
+    eventFeeder.enqueue(secondEvent);
+    await().atMost(5, SECONDS).until(() -> trackedEvents.get(0) != null);
+    await().atMost(5, SECONDS).until(() -> trackedEvents.get(1) != null);
+    assertThat(trackedEvents.get(0), is(firstEvent));
+    assertThat(trackedEvents.get(1), is(secondEvent));
+  }
+
+  @Test(expected = IsClosed.class)
+  public void shouldRejectEventIfClosed() throws Exception {
+    eventFeeder.close();
+    eventFeeder.enqueue(firstEvent);
+  }
+
+  @Test
+  public void ShouldCloseGracefully() throws Exception {
+    EventFeeder<SequenceEvent> eventConsumer =
+        new EventFeeder<>(new SlowInjectingEventConsumer());
+
+    eventConsumer.enqueue(firstEvent);
+    eventConsumer.close();
+    assertThat(trackedEvents.get(0), is(firstEvent));
+  }
+
+  private class InjectingEventConsumer implements Consumer<SequenceEvent> {
+    @Override
+    public void accept(SequenceEvent sequenceEvent) {
+      trackedEvents.add(sequenceEvent);
+    }
+  }
+
+  private class SlowInjectingEventConsumer implements Consumer<SequenceEvent> {
+    @Override
+    public void accept(SequenceEvent sequenceEvent) {
+      try {
+        //Todo better
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+      trackedEvents.add(sequenceEvent);
+    }
+  }
+
+  private class ExceptionalEventConsumer implements Consumer<SequenceEvent> {
+    @Override
+    public void accept(SequenceEvent sequenceEvent) {
+      try {
+        //Todo better
+        Thread.sleep(1000);
```

Comment:
This is can be a good hint that we can probably inject executor to `EventFeeder`.